### PR TITLE
[SIM_FLUTTER-56] [Integrate] Implement user filter by category/type of transaction

### PR DESCRIPTION
**Pull request by jescabillas**
### Links
[Backlog](https://framgiaph.backlog.com/view/SIM_FLUTTER-56)
[Slack](https://sun-philippine.slack.com/archives/C059L7XAXFB/p1689242206394529)

### Definition of done
- [x] User will be able to filter the transaction history by type
- [x] Transaction history list will update accordingly based from transaction type

### Notes
None

### Recordings

https://github.com/framgia/sph-flutter/assets/114911074/006a6080-7c0a-48ae-8a23-1f27163d3100

### CODE REVIEW

**Comment by Dbulawan13** [comment]
Comment:

- [x] Unable to run the application. An error will occur

image


Comment: 1

**Comment by Dbulawan13** [comment]
Comment:

- [x] By default, it should display all the transaction history. 


- [x] By default, filter by category should be blank first


- [x] Only able to display maximum of 10 transactions maybe with this reason the received transaction was not displayed
image

- [x] It display duplicate transactions


- [x] Filter transaction should reset when user visit another account. Or by default it should display first all transactions and filter field should be blank

https://github.com/framgia/sph-flutter/assets/86758474/11d3ffb6-a391-4b76-8b04-5849c6a1e9ad

Comment: 5

**Comment by Dbulawan13** [comment]
Comment:

- [ ] Instead of displaying the text 'All Transactions,' please display it as blank. This will align with the default blank display of the filter by date.


Comment: 1

**Review by Dbulawan13 — DISMISSED** [review]
Approve PR #66 

Working:
- Able to filter category 
   - by All Transaction
   - Credit
   - Dept
   - Transfer 
- All transaction will be selected by default
- All transaction history will be displayed by default
- Once user redirect to another accounts transaction. Filtered category will reset

https://github.com/framgia/sph-flutter/assets/86758474/6a3a9bdf-9970-41c1-a517-981d109405f5

Testing Hours: 2
Comments: 7
Iterations: 3

**Review by Dbulawan13 — DISMISSED** [review]

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 1]
```diff
@@ -26,107 +26,120 @@ class TransactionHistory extends StatelessWidget {
     final arguments =
         ModalRoute.of(context)!.settings.arguments as ScreenArguments;
 
-    return FutureBuilder(
-      future: controller.getTransactions(accountId: arguments.accountId),
-      builder: (context, snapshot) {
-        if (snapshot.connectionState != ConnectionState.done) {
-          return const CircularProgressIndicator();
-        }
+    return Obx(
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: do not wrap FutureBuilder here with Obx

Obx doesnt have an observable child close to it

this was not rebased properly

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 2]
```diff
@@ -26,107 +26,120 @@ class TransactionHistory extends StatelessWidget {
...
+                                ? 'All transactions'
+                                : capitalizeFirstLetter(type.name),
+                          )
+                          .toList(),
+                      selectedValue: controller.selectedTransactionType ==
+                              TransactionTypes.ALL
+                          ? 'All transactions'
+                          : capitalizeFirstLetter(
+                              controller.selectedTransactionType.name,
+                            ),
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: refactor enum `TransactionTypes`
to be similar implementation with `AccountType` enum which is using enhanced enum implementation

use `fromJson` to signify that the static constructor is for the BE data.

File: frontend/lib/src/features/account_details/account_details_page.dart [hunk 3]
```diff
@@ -210,6 +213,8 @@ class AccountDetailsPage extends StatelessWidget {
                           size: const Size(110, 90),
                           radius: 8,
                           onPressed: () {
+                            transactionController
```
Comment (jstephend-sun) [inline frontend/lib/src/features/account_details/account_details_page.dart]: this causes unnecessary double request on the transactions api endpoint

[screen-capture.webm](https://github.com/framgia/sph-flutter/assets/95029803/795502bc-c3dd-467f-aa9d-f8f708a312ad)

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 4]
```diff
@@ -26,107 +26,120 @@ class TransactionHistory extends StatelessWidget {
...
+                  children: [
+                    Dropdown(
+                      labelText: 'Filter transaction by type',
+                      items: TransactionTypes.values
+                          .map(
+                            (type) => type == TransactionTypes.ALL
+                                ? 'All transactions'
+                                : capitalizeFirstLetter(type.name),
+                          )
+                          .toList(),
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: refactor into a function in the enum file, to get the values of the enum as list of string

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 5]
```diff
@@ -26,107 +26,120 @@ class TransactionHistory extends StatelessWidget {
...
+                      onChanged: (value) {
+                        final convertValue =
+                            value.toString() == 'All transactions'
+                                ? 'All'
+                                : value;
+
+                        controller.setSelectedTransactionType =
+                            TransactionTypes.values.firstWhere(
+                          (type) =>
+                              capitalizeFirstLetter(type.name) == convertValue,
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: refactor as static enum constructor

use `fromValue`
contrary from `fromJson`
fromValue signifies that this is from the converted String of the enum

ex.
json = DEPT (which is equal to enum.name)
value = Debit

fromJson("DEPT") returns TransactionType.DEPT
fromValue("Debit") returns TransactionType.DEPT

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #66

review time: 15mins
comments: 5
iteration: 1st review

File: frontend/lib/src/features/transaction_history/transaction_history.dart (hunk 5 above)
Comment (jescabillas) [reply frontend/lib/src/features/transaction_history/transaction_history.dart]: adjusted

File: frontend/lib/src/features/transaction_history/transaction_history.dart (hunk 4 above)
Comment (jescabillas) [reply frontend/lib/src/features/transaction_history/transaction_history.dart]: adjusted

File: frontend/lib/src/features/transaction_history/transaction_history.dart (hunk 2 above)
Comment (jescabillas) [reply frontend/lib/src/features/transaction_history/transaction_history.dart]: adjusted

File: frontend/lib/src/features/transaction_history/transaction_history.dart (hunk 1 above)
Comment (jescabillas) [reply frontend/lib/src/features/transaction_history/transaction_history.dart]: adjusted

File: frontend/lib/src/features/account_details/account_details_page.dart (hunk 3 above)
Comment (jescabillas) [reply frontend/lib/src/features/account_details/account_details_page.dart]: adjusted

**Review by jstephend-sun — APPROVED** [review]
Approved #66

review time: 15mins
comments: 5
iteration: 1st review

**Pull request merged**